core/src/main/scala/chisel3/package.scala

```diff
@@ -210,29 +211,116 @@ package object chisel3 {
+     *   2. If format specifiers specifically target Chisel types - Like Data/Bits -  specific internal methods are called on that argument (like FirrtlFormat)
+     *   3. If format specifers target regular scala data types - java.lang.String.format method is called with specifier and argument to get the formatted string.
+     *
+     * Yet to document all the corener cases which this code tries to cover - but here are few of them
+     * 1. Allow calling %f or %2.2f (or something similar) on Int type data (Int, Short, Long) - this requires doing an explicit cast of the data before calling String.format
+     * 2. For bits - if no format specifier given (%s i.e) call .toPrintable to pick the default one. For regular data - let the String.format handle canonical case of %s too.
+     * Still testing this code - I am pretty sure this can be refactored and made more efficient / readable. WIP.
+     *
+     * /
+     */
+     */
+    def cf(args: Any*): Printable = {
+
+      // Handle literal % 
+      // Takes the part string - 
+      // - this is assumed to not have any format specifiers - already handled / removed before calling this function. 
+      // Only thing present is literal % if any which should ideally be with %%. 
+      // If not - then flag an error. 
+      // Return seq of Printables (either PString or Percent or both - nothing else
+      def percentSplitter(s : String): Seq[Printable] = {
+          if(s.isEmpty) return Seq(PString(""))
+	        val pieces = s.split("%%").toList.flatMap { p =>
+	          if (p.contains('%')) throw new UnknownFormatConversionException("Un-escaped % found")
+	          // Wrap in PString and intersperse the escaped percentages
+	          Seq(Percent, PString(p))
+	        }
+	        if(pieces.isEmpty) Seq(Percent) else pieces.tail  // Don't forget to drop the extra percent we put at the beginning
+        }
+
+      def extractFormatSpecifier(part: String): (Option[String], String) = {
+         // Check if part starts with a format specifier (with % - disambiguate with literal % checking the next character if needed to be %)
+          // In the case of %f specifier there is a chance that we need more information - so capture till the 1st letter (a-zA-Z).
+          // Example cf"This is $val%2.2f here" - parts - Seq("This is ","%2.2f here") - the format specifier here is %2.2f.
+          val idx_of_fmt_str =
+            if (
+              !part.isEmpty() && part.charAt(0) == '%' && (part.size == 1 || (part.size >= 2 && part.charAt(1) != '%'))
+            ) part.indexWhere { _.isLetter }
+            else -1
+
+          val fmt = if (idx_of_fmt_str >= 0) Some(part.substring(0, idx_of_fmt_str + 1)) else None
+
+          val formatRemovedString = part.zipWithIndex.collect { case (p, idx) if idx > idx_of_fmt_str => p }.mkString
+
+          (fmt,formatRemovedString)
```

Comment:
```suggestion
          val endFmtIdx =
            if (part.length > 1 && part(0) == '%' && part(1) != '%') part.indexWhere(_.isLetter)
            else -1
          val (fmt, rest) = part.splitAt(endFmtIdx)
          
          val fmtOpt = if (fmt.nonEmpty) Some(fmt) else None
          (fmtOpt, rest)
```
Overall logic looks good, here's a slight simplification to the condition (check my logic) and the use of `splitAt` also helps.